quickwit/quickwit-metastore/src/metastore/file_backed_metastore/file_backed_metastore_factory.rs

```diff
@@ -44,11 +44,9 @@ use crate::{FileBackedMetastore, Metastore, MetastoreFactory, MetastoreResolverE
 #[derive(Clone)]
 pub struct FileBackedMetastoreFactory {
     storage_resolver: StorageResolver,
-    // We almost never garbage collect the dangling Weak pointers
-    // here. This is judged to not be much of a problem however.
-    //
-    // In a normal run, this cache will contain a single Metastore.
-    cache: Arc<Mutex<HashMap<Uri, Weak<dyn Metastore>>>>,
+    // We never garbage collect unused metastore client instances. This should not be a problem
+    // because during normal use this cache will hold at most a single instance.
+    cache: Arc<Mutex<HashMap<Uri, MetastoreServiceClient>>>,
```

Comment:
If this is true, maybe we should use an `Arc<Mutex<Option<(Uri, MetastoreServiceClient)>>>` and always evict the old client if URIs don't match (and log a warning?)

Reply:
I like the idea and I tested it. Only a few tests are failing if we changed this. We need to also to refactor the factory as now the code between the file backed factory cache and the postgres factory cache is exactly the same.